core/unix/coredump.c

```diff
@@ -0,0 +1,362 @@
+#    define ELF_PROGRAM_HEADER_TYPE Elf64_Phdr
+#    define ELF_SECTION_HEADER_TYPE Elf64_Shdr
+#    define ELF_ADDR Elf64_Addr
+#    define ELF_WORD Elf64_Xword
+#    define ELF_OFF Elf64_Word
+#else
+#    define ELF_HEADER_TYPE Elf32_Ehdr
+#    define ELF_PROGRAM_HEADER_TYPE Elf32_Phdr
+#    define ELF_SECTION_HEADER_TYPE Elf32_Shdr
+#    define ELF_ADDR Elf32_Addr
+#    define ELF_WORD Elf32_Word
+#    define ELF_OFF Elf32_Word
+#endif
+
+DECLARE_CXTSWPROT_VAR(static mutex_t dump_core_lock, INIT_LOCK_FREE(dump_core_lock));
+
+/*
+ * Writes an ELF header to the file. Returns true if the ELF header is written to the core
+ * dump file, false otherwise.
+ */
+static bool
+write_elf_header(DR_PARAM_IN file_t elf_file, DR_PARAM_IN ELF_ADDR entry_point,
+                 DR_PARAM_IN ELF_OFF section_header_table_offset,
+                 DR_PARAM_IN ELF_OFF flags, DR_PARAM_IN ELF_OFF program_header_count,
+                 DR_PARAM_IN ELF_OFF section_header_count,
+                 DR_PARAM_IN ELF_OFF section_string_table_index)
+{
+    ELF_HEADER_TYPE ehdr;
+    ehdr.e_ident[0] = ELFMAG0;
+    ehdr.e_ident[1] = ELFMAG1;
+    ehdr.e_ident[2] = ELFMAG2;
+    ehdr.e_ident[3] = ELFMAG3;
+    ehdr.e_ident[EI_CLASS] = ELFCLASS64;
+    ehdr.e_ident[EI_DATA] = IF_AARCHXX_ELSE(ELFDATA2MSB, ELFDATA2LSB);
+    ehdr.e_ident[EI_VERSION] = EV_CURRENT;
+    ehdr.e_ident[EI_OSABI] = ELFOSABI_LINUX;
+    ehdr.e_ident[EI_ABIVERSION] = 0;
+    ehdr.e_type = ET_CORE;
+    ehdr.e_machine = IF_AARCHXX_ELSE(EM_AARCH64, EM_X86_64);
+    ehdr.e_version = EV_CURRENT;
+    /* This is the memory address of the entry point from where the process starts
+     * executing. */
+    ehdr.e_entry = entry_point;
+    /* Points to the start of the program header table. */
+    ehdr.e_phoff = sizeof(ehdr);
+    /* Points to the start of the section header table. */
+    ehdr.e_shoff = section_header_table_offset;
+    ehdr.e_flags = 0;
+    /* Contains the size of this header */
+    ehdr.e_ehsize = sizeof(ehdr);
+    /* Contains the size of a program header table entry. As explained below, this will
+     * typically be 0x20 (32 bit) or 0x38 (64 bit). */
+    ehdr.e_phentsize = sizeof(ELF_PROGRAM_HEADER_TYPE);
+    /* Contains the number of entries in the program header table. */
+    ehdr.e_phnum = program_header_count;
+    ehdr.e_shentsize = sizeof(ELF_SECTION_HEADER_TYPE);
+    /* Contains the number of entries in the section header table. */
+    ehdr.e_shnum = section_header_count;
+    /* Contains index of the section header table entry that contains the section names.
+     */
+    ehdr.e_shstrndx = section_string_table_index;
+
+    return os_write(elf_file, (void *)&ehdr, sizeof(ELF_HEADER_TYPE)) ==
+        sizeof(ELF_HEADER_TYPE);
+}
+
+/*
+ * Writes a program header to the file. Returns true if the program header is written to
+ * the core dump file, false otherwise.
+ */
+static bool
+write_phdrs(DR_PARAM_IN file_t elf_file, DR_PARAM_IN ELF_WORD type,
```

Comment:
If it writes one, s/phdrs/phdr/?  Or s/phdrs/program_header/ sounds even better.

Reply:
Done.